columnar/src/backend/columnar/columnar_tableam.c

```diff
@@ -898,6 +899,33 @@ columnar_multi_insert(Relation relation, TupleTableSlot **slots, int ntuples,
 
 		uint64 writtenRowNumber = ColumnarWriteRow(writeState, values,
 												   tupleSlot->tts_isnull);
+
+		EState *estate = create_estate_for_relation(relation);
```

Comment:
Can we move some of this objects outside loop?

Reply:
no, enough state is changed that we need to recreate inside the loop, unfortunately.